src/page/slot.rs

```diff
@@ -334,16 +286,161 @@ impl<T, C: cfg::Config> Slot<T, C> {
+impl<T, C> Slot<Option<T>, C>
+where
+    C: cfg::Config,
+{
     fn is_empty(&self) -> bool {
         self.item.with(|item| unsafe { (*item).is_none() })
     }
+
+    /// Insert a value into a slot
+    ///
+    /// We first initialize the state and then insert the pased in value into the slot.
+    #[inline]
+    pub(super) fn insert(&self, value: &mut Option<T>) -> Option<Generation<C>> {
+        debug_assert!(self.is_empty(), "inserted into full slot");
+        debug_assert!(value.is_some(), "inserted twice");
+
+        let gen = self.initialize_state()?;
+        // Set the new value.
+        self.item.with_mut(|item| unsafe {
+            *item = value.take();
+        });
+
+        test_println!("-> inserted at {:?}", gen);
+
+        Some(gen)
+    }
+
+    /// Tries to remove the value in the slot
+    ///
+    /// This method tries to remove the value in the slot. If there are existing references, then
+    /// the slot is marked for removal and the next thread calling either this method or
+    /// `remove_value` will do the work instead.
+    #[inline]
+    pub(super) fn try_remove_value<F: FreeList<C>>(
+        &self,
+        gen: Generation<C>,
+        offset: usize,
+        free: &F,
+    ) -> Option<T> {
+        if self.mark_for_mutation(gen) {
+            None
+        } else {
+            // Otherwise, we can remove the slot now!
+            test_println!("-> remove deferred; can remove now");
+            self.mutate_storage_with_confirmation(gen, offset, free, |item| unsafe {
+                (*item).take()
+            })?
+        }
+    }
+
+    #[inline]
+    pub(super) fn remove_value<F: FreeList<C>>(
+        &self,
+        gen: Generation<C>,
+        offset: usize,
+        free: &F,
+    ) -> Option<T> {
+        self.mutate_storage_with_confirmation(gen, offset, free, |item| unsafe { (*item).take() })?
+    }
+}
+
+impl<T, C> Slot<T, C>
+where
+    T: Default + Clear,
+    C: cfg::Config,
+{
+    pub(in crate::page) fn new(next: usize) -> Self {
+        Self {
+            lifecycle: AtomicUsize::new(0),
+            item: CausalCell::new(T::default()),
+            next: CausalCell::new(next),
+            _cfg: PhantomData,
+        }
+    }
+
+    /// Try to clear this slot's storage
+    ///
+    /// If there are references to this slot, then we mark this slot for clearing and let the last
+    /// thread do the work for us.
+    #[inline]
+    pub(super) fn try_clear_storage<F: FreeList<C>>(
+        &self,
+        gen: Generation<C>,
+        offset: usize,
+        free: &F,
+    ) -> bool {
+        if self.mark_for_mutation(gen) {
+            false
+        } else {
+            // Otherwise, we can remove the slot now!
+            test_println!("-> remove deferred; can remove now");
+            self.mutate_storage_with_confirmation(gen, offset, free, |item| unsafe {
+                (*item).clear()
+            })
```

Comment:
nit: this could just be
```suggestion
            self.clear_storage(gen, offset, free)
```